download: switch to zst compression

aparcar opened the pull request:
The compression is faster to (un)pack files, make use of it.

Also add a new build prerequirement, the `zstd` to (un)pack files.

aparcar committed:
include/download.mk: switch to zst compression

The compression is faster to (un)pack files, make use of it.

Also add a new build prerequirement, the `zstd` to (un)pack files.

Signed-off-by: Paul Spooren <mail@aparcar.org>
[ improve commit title ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

aparcar committed:
treewide: update PKG_MIRROR_HASH to zst

When using zst instead of xz, the hash changes. This commit fixes the
hash for packages and tools in core.

Signed-off-by: Paul Spooren <mail@aparcar.org>

Ansuel committed:
scripts/dl_github_archive: add support for packing zstd archive

Add support for packing .zst archive when creating Github mirror tar.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

aparcar commented:
@Ansuel you typoed `xstd` :)

Ansuel commented:
@aparcar should we limit zstd compression option?

aparcar commented:
> @aparcar should we limit zstd compression option?

I guess you always want ultra or something?

jow- commented:
Why? While Zstd might provide some small gains in unpack speeds or compression ration, it's still comparatively uncommon. It's not that unpacking source archives is the bottleneck during the entire build process.

neheb commented:
@jow- I suggested it as the apk changes are causing tarballs to be generated with different hashes. Such that apk tarballs == zstd and opkg tarballs == xz.

aparcar commented:
Compressing i.e. kernels sources takes its sweet time on my device, eventually it adds up. More importantly are ImageBuilder, SDK et al, so if we switch those to zst we save multiple minutes per build and this source commit here is just a freebie.

neheb commented:
@aparcar @jow- in light of the whole xz mess going on right now, I believe it's a good idea to go forward with this.

Facebook has much more stringent security policies to prevent this from happening.

Ansuel commented:
(this message is just a joke and it's all fake)
Funny coincidence, looks like we were aware of it and we were preparing for it.

robimarko commented:
I agree that its time to do this, especially now that XZ fiasco happened.

neheb commented:
so the issue here is https://github.com/openwrt/openwrt/blob/main/tools/Makefile#L159****

Basically xz gets built before libdeflate and allows it to get unpacked and then used. After this PR it's zstd that must be present. But as zstd currently relies on meson and pkgconfig, those would have to move to core too.

Ansuel commented:
> But as zstd currently relies on meson and pkgconfig, those would have to move to core too.

meson is not already part of core? also do we really need pkgconfig? can't we reference the required lib for ztsd with some args instead of pkgconfig?

Ansuel commented:
If needed we can fix that... So do you confirm meson is already in core?

Ansuel commented:
It is https://github.com/openwrt/openwrt/tree/main/tools/meson

neheb commented:
I'm talking about tools-core in https://github.com/openwrt/openwrt/blob/main/tools/Makefile

Ansuel commented:
Oh ok sorry... zstd only supports meson ?

neheb commented:
regular Make, CMake, and meson. CMake is no go since too many dependencies. The reason I went with meson is because it was easier to make sure the output binaries were statically linked to avoid needing rpath hacks,

robimarko commented:
On Wed, 3 Apr 2024, 00:05 Christian Marangi, ***@***.***>
wrote:

> Oh ok sorry... zstd only supports meson ?
>

Zstd has a bunch of build systems supported, including plain old make which
is the the default one so it can work without meson

> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/14971#issuecomment-2033182706>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKMHMGHJPLOCFRD5BNNL6DY3MTRVAVCNFSM6AAAAABFGTJ3QGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMZTGE4DENZQGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

neheb commented:
Maybe

```
--- a/tools/Makefile
+++ b/tools/Makefile
@@ -157,10 +157,13 @@ $(foreach tool, $(tools-y), $(if $(wildcard $(curdir)/$(tool)/patches),$(eval $(
 $(foreach tool, $(filter-out xz,$(tools-y)), $(eval $(curdir)/$(tool)/compile += $(curdir)/xz/compile))
 
 # make any tool depend on the following to ensure that archives can be unpacked and patched properly
+tools-core += meson
+tools-core += ninja
 tools-core += libdeflate
 tools-core += patch
+tools-core += pkgconfig
 tools-core += tar
-tools-core += xz
+tools-core += zstd
 
 $(foreach tool, $(tools-y), $(eval $(curdir)/$(tool)/compile += $(patsubst %,$(curdir)/%/compile,$(tools-core))))
 tools-y += $(tools-core)
```

Is needed.

Ansuel commented:
Working on it but having some problem with static linking...

Ansuel committed:
tools: zstd: convert to make and drop meson dependency

Convert to make and drop meson dependency since it's not a core tools
and can't depend on advanced build system like cmake or meson.

On top of this make is the official build support and cmake/meson are
supported by 3rd parties.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
tools: make zstd tool core and xz tool non-core

Make zstd tool core and xz tool non-core since zstd is the new default
for compression format.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@neheb what do you think of the conversion? Did I miss anything? Also with meson static build was also triggered with multithread support?

neheb commented:
```
> curl https://patch-diff.githubusercontent.com/raw/openwrt/openwrt/pull/14971.patch | git am
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 27254    0 27254    0     0  61037      0 --:--:-- --:--:-- --:--:-- 60970
Applying: download: switch to zst compression
error: patch failed: include/download.mk:13
error: include/download.mk: patch does not apply
Patch failed at 0001 download: switch to zst compression
hint: Use 'git am --show-current-patch=diff' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
```

Anyway, just make sure ldd reports no host libraries on staging_dir/host/bin/zstd and libzstd.pc looks something like:

```
prefix=/home/mangix/devstuff/openwrt/staging_dir/host
includedir=${prefix}/include
libdir=${prefix}/lib

Name: libzstd
Description: fast lossless compression algorithm library
URL: https://facebook.github.io/zstd/
Version: 1.5.5
Libs: -L${libdir} -lzstd -pthread
Cflags: -I${includedir} -pthread
```

Ansuel commented:
@neheb 

```
 ansuel@Ansuel-XPS  ~/openwrt-ansuel/staging_dir/host   main ± ldd bin/zstd
        linux-vdso.so.1 (0x00007ffe0878d000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007efeac6f6000)
        /lib64/ld-linux-x86-64.so.2 (0x00007efeac9f7000)
 ansuel@Ansuel-XPS  ~/openwrt-ansuel/staging_dir/host   main ± cat lib/pkgconfig/libzstd.pc
prefix=/home/ansuel/openwrt-ansuel/staging_dir/host
includedir=${prefix}/include
libdir=${prefix}/lib

Name: libzstd
Description: fast lossless compression algorithm library
URL: https://facebook.github.io/zstd/
Version: 1.5.5
Libs: -L${libdir} -lzstd -pthread
Cflags: -I${includedir} -pthread
```

neheb commented:
seems good then.

there's a version 1.5.6 btw.

Ansuel commented:
> there's a version 1.5.6 btw.

i think aparcar had a branch for bumping that.

neheb commented:
ah merged.

robimarko commented:
1.5.6 is in master/main, its just that this branch is not rebased

Ansuel commented:
Ok just rebased...

neheb commented:
since we're cleaning up tools/Makefile , why does automake depend on xz?

Ansuel commented:
@neheb did you notice other problem in the tools makefile?

neheb commented:
not really.

tools/zstd might need PKG_CAT:=gunzip .

f00b4r0 commented:
> might want to add HOST_BUILD_PARALLEL:=1 . It's default with meson but not make. No idea if this Makefile is parallel friendly.

Why? It's not hard to test, is it?

```
~/zstd-dev$ make -j48
make[1]: Entering directory '/home/varenet/zstd-dev/lib'
make[1]: Entering directory '/home/varenet/zstd-dev/programs'
creating pkgconfig
[...]
release build completed
make[1]: Leaving directory '/home/varenet/zstd-dev/lib'
==> building with threading support
==> building zstd with .gz compression support
==> no liblzma, building zstd without .xz/.lzma support
==> no liblz4, building zstd without .lz4 support
LINK obj/conf_fab963c4dad570aa62c86cddd1ebdcda/zstd
zstd build completed
```

Let's not unnecessarily slow our build and cripple our buildbots maybe?

aparcar commented:
Looking at https://github.com/openwrt/openwrt/commit/19d3b78304794bc5c87dc74df4c3735715579b5c all source compressions are single threaded and with a change to multithreaded the hash changes anyway, funnily discovered here https://github.com/openwrt/openwrt/pull/15057#issuecomment-2034437153

I'd say zstd is the way to go.

Ansuel commented:
Also other project are switching to it like fwupd

Il Mer 3 Apr 2024, 14:36 Paul Spooren ***@***.***> ha scritto:

> Looking at 19d3b78
> <https://github.com/openwrt/openwrt/commit/19d3b78304794bc5c87dc74df4c3735715579b5c>
> all source compressions are single threaded and with a change to
> multithreaded the hash changes anyway, funnily discovered here #15057
> (comment)
> <https://github.com/openwrt/openwrt/pull/15057#issuecomment-2034437153>
>
> I'd say zstd is the way to go.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/14971#issuecomment-2034492045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQRQOFYXQQZ2KSTSGK3Y3PZV5AVCNFSM6AAAAABFGTJ3QGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDAMZUGQ4TEMBUGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

aparcar commented:
Another argument as the original author of xz states:

> Only the threaded compressor creates files that
    can be decompressed in threaded mode.

via https://git.tukaani.org/?p=xz.git;a=commit;h=6daa4d0ea46a8441f21f609149f3633158bf4704

KanjiMonster commented on tools/Makefile:
```diff
@@ -154,22 +153,22 @@ endif
 tools-core += libdeflate
 tools-core += patch
 tools-core += tar
-tools-core += xz
```
We can't just drop xz, as many upstream tarballs are xz compressed, and we use xz to compress the initramfs if configured. So we still need it.

robimarko replied:
Its not dropped, its just moved to be built later, in the same priority that ZSTD currently is, look for the line 72 diff

KanjiMonster replied:
Misread that part a bit, but AFAICT we still need it here else we become unable to extract .tar.xz tarballs until xz got built, also likewise the above where every tool depends on xz/compile.

robimarko replied:
Well, it seems its still being built first like it used to.
I get your point though, its still required for all of the XZ compressed tarballs that are tool sources

f00b4r0 commented on tools/zstd/Makefile:
```diff
@@ -11,25 +11,32 @@ PKG_LICENSE:=BSD-3-Clause
 
-include $(INCLUDE_DIR)/host-build.mk
-include $(INCLUDE_DIR)/meson.mk
+HOST_BUILD_PARALLEL:=1
```
Most likely unneeded, see https://github.com/openwrt/openwrt/pull/14971#issuecomment-2033829215

f00b4r0 replied:
In fact, it was already removed in the past:
https://github.com/openwrt/openwrt/commit/2e745e9be653626d37248c4f0a88ce9a5f90d183

robimarko commented:
@Ansuel Do you have the issue of git cloned repo tarballs not matching the expected hash?

For example @aparcar updated the hashes but when I do `make tools/check FIXUP=1` hashes change and even weirder even those dont match the expected hash

robimarko committed:
tools: prefer gz or bz2 tarballs

In the light of recent XZ events, and fundamental XZ issues lets work on
moving away from using XZ.

So, use gz compressed tarballs as sources whenever possible.

dwarves only offers bz2 compressed tarballs, so use those as size
difference is minor compared to XZ.

Signed-off-by: Robert Marko <robimarko@gmail.com>

dwarves

neheb commented:
PR seems broken:

```
. /home/mangix/devstuff/openwrt/include/shell.sh;  /home/mangix/devstuff/openwrt/dl/libdeflate-1.19.zst | tar -C /home/mangix/devstuff/openwrt/build_dir/host/libdeflate-1.19/.. -xf -
bash: line 1: /home/mangix/devstuff/openwrt/dl/libdeflate-1.19.zst: Permission denied
tar: This does not look like a tar archive
tar: Exiting with failure status due to previous errors
make[3]: *** [Makefile:41: /home/mangix/devstuff/openwrt/build_dir/host/libdeflate-1.19/.preparedcbad0a9221a7d9cb98bd08643a0126a3_6664517399ebbbc92a37c5bb081b5c53] Error 2
```

neheb committed:
tools: rework tools-core dependecies

ZSTD and libdeflate do not depend on SED nor flock, so instead of the whole
for loop that will filter 2 out of 4 core packages just specify that patch
and tar depend on sed explicitly.

ZSTD now depends on libdeflate since libdeflate-gzip will then be used to
unpack ZSTD as well as most tool archives.

Signed-off-by: Rosen Penev <rosenp@gmail.com>
Signed-off-by: Robert Marko <robimarko@gmail.com>

neheb commented:
I got these two commits:

https://github.com/neheb/openwrt/commit/c757726e4e822b8b5011d53ed411845bd1a1d5f9
https://github.com/neheb/openwrt/commit/856733e0c86c7bcff31292147fb6571bc5ca7f81

to get me this far. That sed and flock stuff doesn't seem to be needed by everything.

neheb commented on include/download.mk:
```diff
@@ -13,7 +13,7 @@ ifdef PKG_SOURCE_VERSION
   endif
   PKG_SOURCE_SUBDIR ?= $(PKG_NAME)-$(PKG_VERSION)
-  PKG_SOURCE ?= $(PKG_SOURCE_SUBDIR).tar.xz
+  PKG_SOURCE ?= $(PKG_SOURCE_SUBDIR).zst
```
.tar.zst?

robimarko replied:
Yeah, exactly as otherwise detection will break and ZSTD tarballs are not extracted

aparcar replied:
Feel free to change this. I saw .zst at other places so I thought it's the way to go.

Ansuel replied:
I also notice for zst .tar is not used... Gnu tar checks the extension to detect a tar archive?

robimarko replied:
No idea, but without piping it via zstdcat it wont get extracted.

neheb commented:
yeah that fixed it.

neheb commented:
@robimarko you broke tools/m4. PKG_CAT needs to be removed.

neheb commented:
@Ansuel that zstd conversion looks like no go:

```
checking for libzstd >= 1.4.0... no
configure: error: --with-zstd was given, but pkgconfig/libzstd.pc is not found
make[5]: *** [Makefile:3018: configure-bfd] Error 1
make[5]: Leaving directory '/home/mangix/devstuff/openwrt/build_dir/toolchain-aarch64_generic_gcc-12.3.0_musl/binutils-2.41'
make[4]: *** [Makefile:1005: all] Error 2
make[4]: Leaving directory '/home/mangix/devstuff/openwrt/build_dir/toolchain-aarch64_generic_gcc-12.3.0_musl/binutils-2.41'
make[3]: *** [Makefile:116: /home/mangix/devstuff/openwrt/build_dir/toolchain-aarch64_generic_gcc-12.3.0_musl/binutils-2.41/.built] Error 2
make[3]: Leaving directory '/home/mangix/devstuff/openwrt/toolchain/binutils'

```

robimarko commented:
@neheb Yeah, we are having a bit of a chicken and egg issue with libdeflate

robimarko committed:
tools: libdeflate: fetch source as tarball

libdeflate is currently intentionally being fetched via GIT.

However, with the move to using ZSTD to compress the cloned GIT repo
tarballs it means that we would first need to compile ZSTD.
But that means that we need to be able to unpack gzipped tarballs first
which we currently do by using libdeflate-gzip.
So, in order to do so lets fetch libdeflate as a tarball, use gzip to
extract it and then use libdeflate as regular for all other tools.

Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
This needs more refinement but now host tools will build.
In the end I opted to fetch libdeflate as tarball and unpack it with gunzip thanks to @neheb so it can be built first

aparcar commented:
@jow- You opposed this change via a comment but in the IRC chat you seemed okay to do the switch? Just want to be sure we're on the same page before moving forward

robimarko commented:
@neheb ZSTD conversion fixed up, `PREFIX` was not being passed so `.pc` wasn't updated

neheb commented on tools/Makefile:
```diff
@@ -69,7 +69,7 @@ tools-y += sstrip
 tools-y += util-linux
 tools-y += zip
 tools-y += zlib
-tools-y += zstd
```
Not sorted

robimarko commented:
@Ansuel This is just a collection that "works" now, it can be improved for sure.

Thing is that I dont trust XZ anymore, who cares about couple of MB compared to that situation

Ansuel commented:
@robimarko can you check this version? https://github.com/Ansuel/openwrt/tree/zstd-source

Seems much cleaner to me... I dropped the xz commit tho but that can be cherrypicked.

f00b4r0 commented on tools/zstd/Makefile:
```diff
@@ -11,25 +11,33 @@ PKG_LICENSE:=BSD-3-Clause
-	-Dlzma=disabled \
-	-Dlz4=disabled
+HOST_MAKE_FLAGS += \
+	ZSTD_LEGACY_SUPPORT=7 \
```
BTW, why are we doing this?
From https://github.com/facebook/zstd/blob/dev/lib/README.md#modular-build
> By default, this build macro is set as ZSTD_LEGACY_SUPPORT=5

robimarko replied:
Current meson build does the same, I guess its just ported straight from meson options

f00b4r0 replied:
so that means it can probably be dropped then, like the non-parallel build :)
I think we really should avoid diverging from upstream defaults unless we have a very good technical reason for doing so, imho.

robimarko replied:
I agree, the ZSTD_LEGACY_SUPPORT=7 seems to appeared during 1.5.2 update and meson conversion.

neheb replied:
I did this as legacy support is not needed. I don’t remember if it shaved a little binary space.

f00b4r0 replied:
I don't think "shaving a little binary space" is very relevant here. Perfect compatibility with upstream defaults seems far more important imho.

Ansuel replied:
Honestly that option just disables support for old incompatible formats... Something that should never happen since nobody should generate stuff from zstd 5.0.

For upstream they use 5 to support ancient archive but this is not the case on OpenWRT... Honestly I would even set it to 0 considering how new the zstd adoption is...

We would use it only for our mirrors generated tar so the incompatibility problem doesn't apply. But I'm open on this... It's really just taste.

f00b4r0 replied:
it's disabling the ability to *read* these older formats as well, meaning the ability to read potentially older archives.

But again the more important point is that we're diverging from vanilla upstream and I think this is something we should really avoid for core tools. It means we'd be using a potentially less tested (covered) variant that's special to *us* only, and I don't think it's a very good idea. My 2c.

robimarko replied:
I kind of have to agree with @f00b4r0 for host tools it doesn't really make sense to save on couple of kB in this case.

Ansuel commented:
An alternative is to use .tar directly... That removes the need for the decompress CMD... (But gzip is a prerew so it's always available.

Anyway declaring the CMD in the makefile after host-build.mk overwrite the one in unpack. (I tested this)

robimarko commented on tools/zstd/Makefile:
```diff
@@ -11,25 +11,32 @@ PKG_LICENSE:=BSD-3-Clause
 
-include $(INCLUDE_DIR)/host-build.mk
-include $(INCLUDE_DIR)/meson.mk
+HOST_BUILD_PARALLEL:=1
```
I am not sure that its the default since host-build.mk does not set it.
Also, all other tools set it to parallel build

f00b4r0 replied:
ah ok, I'm confused then. So it doesn't hurt to leave it there. Sorry for the noise.

neheb commented:
@Ansuel PKG_CAT is considered legacy. Might as well use it.

Ansuel commented:
> @Ansuel PKG_CAT is considered legacy. Might as well use it.

I prefer decompress than adding pkgcat to a random value just to trigger the legacy path

neheb commented on tools/Makefile:
```diff
@@ -164,9 +164,10 @@ tools-core += zstd
+# make most core tools depend on sed and flock
+$(curdir)/patch/compile += $(curdir)/sed/compile
+$(curdir)/tar/compile += $(curdir)/sed/compile
+$(curdir)/zstd/compile += $(curdir)/libdeflate/compile $(curdir)/flock/compile
```
I'm curious. Is flock needed?

robimarko replied:
Well, now this interested me so I tried building it in a simple Alpine docker with just make, gcc and g++ and it builds without flock.

neheb replied:
well, busybox includes flock.

neheb replied:
Ah this is probably for macOS.

robimarko replied:
I doubt ZSTD needs flock, removing the busybox applet symlink still doesnt break building

neheb replied:
Introduced with https://github.com/openwrt/openwrt/pull/14971/commits/23ba9fd813849dc2c8717374710bfca77ead2179 . xz might need flock. zstd does not.

robimarko commented:
> ```
> This branch has conflicts that must be resolved
> ```

Its the recent PKG_HASH fixup, will rebase in the morning

aparcar commented on tools/dwarves/Makefile:
```diff
@@ -6,9 +6,9 @@ PKG_NAME:=dwarves
 
-PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.xz
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.bz2
 PKG_SOURCE_URL:=https://fedorapeople.org/~acme/dwarves/
-PKG_HASH:=ad4c08339850e404609e2808012580b7e98366d2b91054bb93fe6dca94651fb4
+PKG_HASH:=1d8c9a1c2d42e06cc121a70a39c4f621fd28f15c476ed1b7c7b226f41fdd32df
 
```
How about using git and create our own tarballs based on that instead of relying their release? In light of xz fellow dev @dangowrt suggested to use overall more git based sources instead of tarballs, since it's harder to tamper with them. If we change hashes here, might as well switch to git.

```
PKG_SOURCE_PROTO:=git
PKG_SOURCE_URL:=https://github.com/acmel/dwarves.git
PKG_SOURCE_VERSION:=922085e3cff7e976cbb97b2e15c7f866956728ae
PKG_MIRROR_HASH:=xxx
```

I'm not familiar with the performance of bzip2 but from my understanding it's single threaded and didn't see updates in the last 5 ish years.

robimarko replied:
Well, with move to ZSTD this should be feasible without introducing a large performance penalty as XZ compression was rather slow.
It will still be slower than downloading a tarball.

robimarko replied:
I just remembered that for dwarves we can move the source to https://git.kernel.org/pub/scm/devel/pahole/pahole.git/

Which does serve GZ tarballs

f00b4r0 replied:
> How about using git and create our own tarballs based on that instead of relying their release? In light of xz fellow dev @dangowrt suggested to use overall more git based sources instead of tarballs, since it's harder to tamper with them. If we change hashes here, might as well switch to git.

AIUI some members have expressed concerns about doing that, mentioning speed and mirror support among other issues.

> I'm not familiar with the performance of bzip2 but from my understanding it's single threaded and didn't see updates in the last 5 ish years.

Yes, bzip2 has been around for 25 years, it is *very* stable. There are parallel implementations too, should we really need this (although for decompression purposes, I don't know if it makes that much of a difference).

robimarko replied:
I will switch it back to the kernel.org repo, it was switched to the Fedora mirror for 1.23 update.

aparcar replied:
> It will still be slower than downloading a tarball.

Just for clarification, whenever our buildbots download something, they upload the content of `dl/` to sources.openwrt.org, which is CDN'ed.

If the protocol is git, `download.mk` will first try to download the tarball from sources.openwrt.org and only then does a git clone. We're really only cloning the git repository once (via buildbots) and then never again.

> Which does serve GZ tarballs

Cool!

> AIUI some members have expressed concerns about doing that, mentioning speed and mirror support among other issues.

Yes, mirror support is certainly worsened but then again some stuff is only ever available on sources.openwrt.org so it will never go away. Maybe we could look into have more people mirror it? After all it's behind a `fastly`.

> Yes, bzip2 has been around for 25 years, it is very stable. There are parallel implementations too, should we really need this (although for decompression purposes, I don't know if it makes that much of a difference).

While compressing happens only once (see above), decompression happens like all the time so based on [this](https://linuxreviews.org/Comparison_of_Compression_Algorithms#Decompressing_The_Linux_Kernel) the difference is quite big. I know dwarves is not the linux kernel but 8x faster seems like a good argument.

f00b4r0 replied:
@robimarko you're back to using fedorapeople instead of kernel.org, is this what you intended?

robimarko replied:
Yes, as the snapshot feature that provide GZ tarballs on the kernel.org repo does not include git submodules so libbpf is missing and dwarves/pahole wont compile obviously.

aparcar commented:
> Why? While Zstd might provide some small gains in unpack speeds or compression ration, it's still comparatively uncommon. It's not that unpacking source archives is the bottleneck during the entire build process.

https://oftc.irclog.whitequark.org/openwrt-devel/2024-04-03#33074511;

So re-reading this @jow- already approved the change.

Ansuel committed:
include/download.mk: handle .gitattributes rules on rawgit method

This fix a long lasting bug/inconsistency with rawgit method and
dl_github_archive script.

The dl_github_archive script works by using the github generated tar.gz
instead of cloning and checkout and the tar.gz is generated by using git
archive command that parse and apply .gitattributes rules.

rawgit command never handled .gitattributes and instead made a simple git
clone and checkout causing the inconsistency.

To fix the inconsistency, add extra steps to call git archive command
and generate an intermediate tar to apply .gitattributes rules.

Also for git log format, Github shorthash length is 8 instead of the
default 7, also apply this locally for each cloned repo to produce
consistent tar.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
tools: zstd: override max safe compression level

ZTSD limits the safe compression level to a max of 19 as 20 to 22 cause
increased RAM usage. Higher levels require --ultra arg passed.

There isn't currently a way to set --ultra using ENV options similar to
ZSTD_CLEVEL and ZSTD_CLEVEL is limited to 19.

To fix this, we can increase the max safe compression level by providing
a custom ZSTDCLI_CLEVEL_MAX value with CFLAGS.

The max safe level is increased to 20.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

systemcrash commented:
@ansuel re e7235157ff03511dde976cbcd8b662c06d514f48 - overwrite or override? Override seems more appropriate here.

robimarko committed:
tools: zstd: dont override ZSTD_LEGACY_SUPPORT

We dont really have a reason to deviate from the upstream default for
ZSTD_LEGACY_SUPPORT value of 5, as it will save a bit of space but
prevent decompressing data compressed with legacy ZSTD versions.

Signed-off-by: Robert Marko <robimarko@gmail.com>

f00b4r0 commented:
> @Ansuel re [e723515](https://github.com/openwrt/openwrt/commit/e7235157ff03511dde976cbcd8b662c06d514f48) - overwrite or override? Override seems more appropriate here.

Good catch. Agree that "override" is more better :)

robimarko commented:
Its because sed depends on it and technically sed is supposed to build first, but we then add flock and zstd as its dependencies and then libdeflate to zstd

neheb commented:
Does it actually rely on flock?

robimarko commented:
On Fri, 5 Apr 2024, 23:26 Rosen Penev, ***@***.***> wrote:

> Does it actually rely on flock?
>

Probably need to dig through the history but I have a feeling that its not
a direct dependecy of sed but rather just to make sure flock gets built
first as it is used by our buildroot for prevent multiple acess etc.

—
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/14971#issuecomment-2040646028>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKMHMB4QMGQ3TJZDLJSGMDY34JHXAVCNFSM6AAAAABFGTJ3QGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDANBQGY2DMMBSHA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

neheb commented:
Yeah.

f00b4r0 reviewed (approved):
LGTM. Let's ship it!

robimarko commented:
Thanks! Rebased on top of main and merged!

robimarko commented:
Lets see what breaks

stangri commented:
Is this going to be backported to 23.05? Currently this led to cherry-picks of packages from snapshots not building for 23.05.

aparcar commented:
Yea that's very unfortunate and I'm sorry for the inconveniences! I don't think a backport makes sense and it drastically changes the build system and has nothing to do in a point backport

stangri commented:
@aparcar thanks for your prompt reply. Out of the packages I still maintain only one needs to be built from external sources and I do not foresee any updates to it until the end of the year, so not a big deal for me.

stangri commented:
@aparcar Paul, like I said it doesn't affect me personally, but maybe until 23.05 is EOLed, you can introduce a transitional variable like `PKG_MIRROR_HASH_XZ` so that it can be set to the hash of the tar.xz archive and that the same Makefile (containing two hashes) can be used for both snapshots and 23.05? I don't want to use the word nightmare again, but if many non-core packages need to be backported to 23.05 for whatever reasons, it might save some time doing just the cherry-picks for 23.05 instead of having to do separate/additional commit with the appropriate hash instead of/in addition to the cherry-pick.

Just my 2 cents.
